### ISSUE TITLE
Segmentation fault in mod_wsgi with WSGIDaemonProcess

### ISSUE THREAD

**Issue by dims** [GrahamDumpleton/mod_wsgi#94]
Hi,

I can reliably create the following problem seen in Apache2 error logs snippet below. The simplest example for mod_wsgi is enough to recreate this situation with at least a couple of versions of apache (2.4.7 and 2.4.10) that i have tried this on. 

The sequence to recreate is as follows:
1) get the sample up and running, make sure you can see url using your browser.
2) disable the site and the module
```
apachectl stop
a2dissite 001-myweb.conf
a2dismod wsgi
```
3) start apache, enable the wsgi module and the site
```
apachectl start
a2enmod wsgi
a2ensite 001-myweb
```
4) gracefully restart apache
```
apachectl graceful
```
5) see the infinite loop in the logs

Problem seems to stem from the "hack" to smartly side step the case where wsgi_hook_init is called twice (during apache start):
https://github.com/GrahamDumpleton/mod_wsgi/blob/develop/src/server/mod_wsgi.c#L12307

wsgi_hook_child_init is supposed to be called after wsgi_hook_init and uses the information in wsgi_server_config. Problem is that when graceful is triggered, the mod_wsgi.so is unloaded and reloaded BUT wsgi_hook_init is called just once! and when wsgi_hook_child_init is called it fails trying to reference information in wsgi_server_config as it is NULL.

Note that the Apache wiki seems to warn about the problem with this "hack":
https://wiki.apache.org/httpd/ModuleLife

Thanks,
Dims

==== Sample Configuration ====
https://gist.github.com/dims/4dfbac57efabb7631da4#file-001-myweb-conf

==== Sample python ====
https://gist.github.com/dims/4dfbac57efabb7631da4#file-code-py

==== Error Logs ====
[Fri Sep 18 16:30:28.942414 2015] [mpm_event:notice] [pid 17562:tid 140163774752640] AH00489: Apache/2.4.10 (Ubuntu) configured -- resuming normal operations
[Fri Sep 18 16:30:28.945243 2015] [core:notice] [pid 17562:tid 140163774752640] AH00094: Command line: '/usr/sbin/apache2'
[Fri Sep 18 16:30:36.291705 2015] [mpm_event:notice] [pid 17562:tid 140163774752640] AH00493: SIGUSR1 received.  Doing graceful restart
AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 127.0.0.1. Set the 'ServerName' directive globally to suppress this message
[Fri Sep 18 16:30:36.299026 2015] [mpm_event:notice] [pid 17562:tid 140163774752640] AH00489: Apache/2.4.10 (Ubuntu) configured -- resuming normal operations
[Fri Sep 18 16:30:36.299233 2015] [core:notice] [pid 17562:tid 140163774752640] AH00094: Command line: '/usr/sbin/apache2'
[Fri Sep 18 16:30:37.301181 2015] [core:notice] [pid 17562:tid 140163774752640] AH00052: child pid 17645 exit signal Segmentation fault (11)
[Fri Sep 18 16:30:37.301418 2015] [core:error] [pid 17562:tid 140163774752640] AH00546: no record of generation 0 of exiting child 17645
[Fri Sep 18 16:30:37.301656 2015] [core:notice] [pid 17562:tid 140163774752640] AH00052: child pid 17646 exit signal Segmentation fault (11)
[Fri Sep 18 16:30:37.301771 2015] [core:error] [pid 17562:tid 140163774752640] AH00546: no record of generation 0 of exiting child 17646
[Fri Sep 18 16:30:38.304622 2015] [core:notice] [pid 17562:tid 140163774752640] AH00052: child pid 17647 exit signal Segmentation fault (11)
[Fri Sep 18 16:30:38.304866 2015] [core:error] [pid 17562:tid 140163774752640] AH00546: no record of generation 0 of exiting child 17647
[Fri Sep 18 16:30:39.307236 2015] [core:notice] [pid 17562:tid 140163774752640] AH00052: child pid 17648 exit signal Segmentation fault (11)
[Fri Sep 18 16:30:39.307462 2015] [core:error] [pid 17562:tid 140163774752640] AH00546: no record of generation 0 of exiting child 17648
[Fri Sep 18 16:30:40.309109 2015] [core:notice] [pid 17562:tid 140163774752640] AH00052: child pid 17649 exit signal Segmentation fault (11)
[Fri Sep 18 16:30:40.309389 2015] [core:error] [pid 17562:tid 140163774752640] AH00546: no record of generation 0 of exiting child 17649

**Comment by paboldin** [comment]
A solution is maybe to check for ap_scoreboard_image->global->running_generation and if it is != 0 then doing module initialization from the first pre_/post_config call.

Going to check if this will work.

**Comment by paboldin** [comment]
```(gdb) p ap_scoreboard_image->global->running_generation 
$1 = 0
(gdb) c
Continuing.

# (running service apache2 restart in a different session)
Program received signal SIGUSR1, User defined signal 1.
0x00007f9b6568dd83 in __select_nocancel () at ../sysdeps/unix/syscall-template.S:81
81	in ../sysdeps/unix/syscall-template.S
(gdb) c
Continuing.

0x00007f9b6568dd83 in __select_nocancel () at ../sysdeps/unix/syscall-template.S:81
81	in ../sysdeps/unix/syscall-template.S
(gdb) p ap_scoreboard_image->global->running_generation 
$2 = 1```

**Comment by dims** [comment]
Nice work @paboldin thanks!

**Comment by GrahamDumpleton** [comment]
FWIW, the original code causes no problem that I can see/create on MacOS X with Apache shipped by Apple. I am wary at this point of making a change as described as it could then stop things from working properly on MacOS X Apache. Need to understand more about the underlying reason and confirm whether this is caused just by Debian/Ubuntu Apache and whether those Linux distributions have screwed around with Apache code and so caused it.

**Comment by GrahamDumpleton** [comment]
Adding debugging, what happens on MacOS X for start up is:

```
mod_wsgi (pid=90732): Module initialisation hook for mod_wsgi called.
mod_wsgi (pid=90732): Module initialisation flag for mod_wsgi is 0.
...
mod_wsgi (pid=90732): Module initialisation hook for mod_wsgi called.
mod_wsgi (pid=90732): Module initialisation flag for mod_wsgi is 1.
```

Then on graceful restart, triggered by SIGUSR1 as using 'graceful' on 'apachectl' on MacOS X doesn't actually do a graceful, one gets:

```
mod_wsgi (pid=90732): Module initialisation hook for mod_wsgi called for Apache generation 1.
mod_wsgi (pid=90732): Module initialisation flag for mod_wsgi is 1.
```

So it is true that on a graceful restart it is only called once, but on MacOS X, the 'wsgi_init' data key in that case carries through as already being set meaning that initialisation will actually occur as required.

For the code to fail then would imply that on your Linux system, the Apache instance is not doing what Apache has always done, which is that the 'wsgi_init' data key would survive the graceful restart. That data key is attached to the process pool, so the memory for it should still be valid and the record should not be cleared.

Since I can't replicate this, I need to see the output on your systems from adding debugging like the following.

```
    if (!ap_scoreboard_image) {
        ap_log_error(APLOG_MARK, APLOG_CRIT, 0, NULL,
                "mod_wsgi (pid=%d): Module initialisation hook for "
                "mod_wsgi called.", getpid());
    }
    else {
        ap_log_error(APLOG_MARK, APLOG_CRIT, 0, NULL,
                "mod_wsgi (pid=%d): Module initialisation hook for "
                "mod_wsgi called for Apache generation %d.", getpid(),
                ap_scoreboard_image->global->running_generation);
    }

    userdata_key = "wsgi_init";

    apr_pool_userdata_get(&data, userdata_key, s->process->pool);

    ap_log_error(APLOG_MARK, APLOG_CRIT, 0, NULL,
            "mod_wsgi (pid=%d): Module initialisation flag for "
            "mod_wsgi is %d.", getpid(), !!data);

    if (!data) {
        apr_pool_userdata_set((const void *)1, userdata_key,
                              apr_pool_cleanup_null, s->process->pool);
        return OK;
    }
```

I am using prefork MPM, but I don't believe that should make any difference and should not be different to event MPM.

**Comment by GrahamDumpleton** [comment]
So I should read things properly. The issue isn't necessarily so much the graceful restart, but the fact that the module is being loaded into Apache for the first time when doing a graceful restart, as opposed to the module only being enabled when Apache is stopped.

**Comment by GrahamDumpleton** [comment]
Variation on fix added to develop branch for further testing. Fix differs in ensure that the wsgi_init data key is still set to be consistent.

**Comment by GrahamDumpleton** [comment]
Fix released with 4.4.15. Notify if still causes issues.
